Add three new APIs to microsoft.resourcehealth

wengjn opened the pull request:
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

Commit:
Add three new APIs to microsoft.resourcehealth

AutorestCI commented:
# Automation for azure-sdk-for-js

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-java

azuresdkci commented:
Can one of the admins verify this patch?

Commit:
fix the schema for new apis

reuse the old schema instead of creating new ones

Commit:
fix schema another error

KrisBash commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/stable/2015-01-01/resourcehealth.json:
```diff
@@ -221,6 +222,154 @@
         }
       }
     },
+    "/{resourceUri}/providers/Microsoft.ResourceHealth/child/availabilityStatuses/current": {
```
The "child" token in these paths is out of spec with ARM contract. If this a singleton - and there will not be many child instances-, this is more appropriate:
`/{resourceUri}/providers/Microsoft.ResourceHealth/childAvailabilityStatuses/current`

KrisBash commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/stable/2015-01-01/resourcehealth.json:
```diff
@@ -221,6 +222,154 @@
+        }
+      }
+    },
+    "/{resourceUri}/providers/Microsoft.ResourceHealth/child/availabilityStatuses": {
```
Same as above. I'd recommend /childAvailabilityStatuses

KrisBash commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/stable/2015-01-01/resourcehealth.json:
```diff
@@ -221,6 +222,154 @@
+        }
+      }
+    },
+    "/{resourceUri}/providers/Microsoft.ResourceHealth/children": {
```
I can ask for a second opinion, but I'd prefer the resource type name of "childResources" over "children." While "children" is grammatically plural, it feels inconsistent.

KrisBash commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/stable/2015-01-01/resourcehealth.json:
```diff
@@ -351,6 +500,10 @@
               "description": "Timestamp for when the health was last checked. ",
               "format": "date-time"
             },
+            "isArmResource": {
```
One thing to consider is if this would be more future-proof as enum, such as resourceType or resourceManager with values of ARM or other. If you may want to distinguish between types of non-ARM resources in the future, it's more trivial to expand a string enum.

wengjn replied:
Thanks very much for your suggestion. This flag would just stay as boolean as it only indicate to the client that whether to show further details for child resource. If it is arm type resource, we would be able to get more details. If not, client/caller would not be able to get child health status. That's it. As of now, I don't see necessary of making any future-proof on this property. :)

KrisBash commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/stable/2015-01-01/resourcehealth.json:
```diff
@@ -221,6 +222,154 @@
+          "200": {
+            "description": "The body contains the list of the historical availability statuses for a single child resource",
+            "schema": {
+              "$ref": "#/definitions/childAvailabilityStatusListResult"
```
I'm not seeing this definition in the swagger

wengjn replied:
Thanks for looking into this. this is the old commit which I have changed. I'm not adding new schema. I'm reusing the existing ones that we have.

KrisBash commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/stable/2015-01-01/resourcehealth.json:
```diff
@@ -221,6 +222,154 @@
+          "200": {
+            "description": "The body contains the list of the children's current availability statuses for a single resource which contains children",
+            "schema": {
+              "$ref": "#/definitions/ChildrenListResult"
```
I'm not seeing this definition in the swagger

wengjn replied:
Thanks for looking into this. this is the old commit which I have changed. I'm not adding new schema. I'm reusing the existing ones that we have.

KrisBash commented on specification/resourcehealth/resource-manager/Microsoft.ResourceHealth/stable/2015-01-01/resourcehealth.json:
```diff
@@ -221,6 +222,154 @@
+          "200": {
+            "description": "The body contains the current availability status for a single resource",
+            "schema": {
+              "$ref": "#/definitions/childAvailabilityStatus"
```
I'm not seeing this definition in the swagger

wengjn replied:
Thanks for looking into this. this is the old commit which I have changed. I'm not adding new schema. I'm reusing the existing ones that we have.

KrisBash reviewed (changes requested):
Changes needed from ARM

KrisBash commented:
@wengjn feel free to set up a meeting with ARM API review if you'd like to discuss.

ravbhatnagar commented:
@wengjn - please let us know when this is ready for re-review by ARM or feel free to set up a meeting.

Commit:
Change resource type name as suggested from ARM review

Change resource type name as suggested from ARM review: childAvailabilityStatuses and childResources

wengjn commented:
This is updated. All comments have been resolved. Thx

KrisBash reviewed (approved):
LGTM after changes. Thanks
